### PR TITLE
stage collection: do not collect .gitignored dvcfiles

### LINKED ISSUE #5244: stage collection: ignore git-ignored directories?
`dvc status` in our repo is terribly slow (~6s), and I think a project of a reasonable size of our users are slow as well, as we might be traversing through very dense directories (eg: `node_modules`/`venv`, and user's dvc-tracked data).

We have discussed before to get rid of `.dvcignore`, but I see it to be a different thing, useful in, for example, ignoring `.DS_Store`/`logs`, etc in dvc-tracked artefacts.

We could opt for a simple solution and just ignore the git-ignored dvcfiles and directories and never traverse them. In addition to this, as we also add user's data in the `gitignore`, we can also get significant speedups just by doing this.

Also, design-wise, we won't need to mix `gitignore`s and `dvcignore`s in the `tree`s.

##### Before
```console
  Time (mean ± σ):      4.240 s ±  0.192 s    [User: 3.535 s, System: 0.474 s]
  Range (min … max):    4.141 s …  4.781 s    10 runs
```

#### After
```console
Benchmark #1: dvc status with dulwich cached
  Time (mean ± σ):     512.1 ms ±  73.2 ms    [User: 387.6 ms, System: 43.3 ms]
  Range (min … max):   436.2 ms … 704.0 ms    10 runs
```

Also, note that we can change this behaviour in 2.0 release, as this could be considered a breaking change.

### PR DESCRIPTION
**Pull request by skshetry**
Closes #5244 

A few fixes has been made along the way, including:
* better error message if the dvc.yaml and .dvc files are dvc-ignored.
  ```console
  $ echo "dvc.yaml" >> .dvcignore
  $ dvc repro
  ERROR: 'dvc.yaml' is dvc-ignored. 
  ```
* better error message if the dvc.yaml and .dvc files are git-ignored.
  ```console
  $ echo "dvc.yaml" >> .gitignore
  $ dvc repro
  ERROR: 'dvc.yaml' is git-ignored.
  ```

An open question is, what should we do about lockfiles if they are git-ignored or dvc-ignored?
Right now, we assume the lockfile as empty if they are dvc-ignored and we don't care about git-ignores
for the lockfiles. I feel that this behaviour is good, but if we want them to be consistent, please do tell me.

Another thing to keep in mind is, since the implementation depends on `isinstance(repo.tree, LocalTree)`, if we start using LocalTree for `get/list/import`, it'll break, though I don't think it should as we better apply it to only the user's workspace (IIRC `dvcx` uses this mode).
Though, this might be handled by a state (eg: `repo.is_external_repo` or something like that).

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/repo/stage.py [hunk 1]
```diff
@@ -371,13 +372,33 @@ def collect_repo(self, onerror: Callable[[str, Exception], None] = None):
...
-        stages = []
+        scm = self.repo.scm
+        sep = os.sep
         outs: Set[str] = set()
+
+        is_local_tree = isinstance(self.tree, LocalTree)
+
+        def is_ignored(path):
+            # apply only for the local tree
+            return is_local_tree and scm.is_ignored(path)
```
Comment (skshetry) [inline dvc/repo/stage.py:385]: This check might not work everywhere, especially in `for_write=True` scenarios in `dvcx`. But as we are adjusting our APIs in `ExternalRepo`, `Repo.clone()` should set some kind of state to workaround that.

**Review by efiop — APPROVED** [review]

**Comment by pared** [comment]
> An open question is, what should we do about lockfiles if they are git-ignored or dvc-ignored?

I think users should be able to ignore lockfile. Dvc is able (to some extent) to work without it, so removing this possibility would limit `dvc`'s functionality. And I guess one can come with some use cases where this is actually a feature. For example any educational material that's purpose is to show how running the pipeline looks like.

**Review by pmrowla — APPROVED** [review]

**Comment by pmrowla** [comment]
> > An open question is, what should we do about lockfiles if they are git-ignored or dvc-ignored?
> 
> I think users should be able to ignore lockfile. Dvc is able (to some extent) to work without it, so removing this possibility would limit `dvc`'s functionality. And I guess one can come with some use cases where this is actually a feature. For example any educational material that's purpose is to show how running the pipeline looks like.

One thing to keep in mind is that experiments runs in temp directories (and on remote machines in the future) have to git commit `dvc.lock` in order for us to retrieve the experiment result. We can (and already do) force-add gitignored lock files internally for these types of runs, and on the user's local `exp apply` the dvc.lock file would no longer be staged/committed like normal. But this does seem like the kind of thing that may lead to unexpected behavior in the future and it may be safer for us to forbid it?

**Comment by pared** [comment]
@pmrowla Fair point, I didn't consider experiments in my thought process. In that case it might be better to just require it.

**Pull request merged**